src/circuit/derive_nullifier.rs

```diff
@@ -0,0 +1,112 @@
+//! Derive nullifier logic for the Orchard circuit.
+
+use crate::constants::nullifier_l::nullifier_l;
+use halo2_gadgets::utilities::cond_swap::CondSwapChip;
+use halo2_proofs::circuit::AssignedCell;
+use pasta_curves::pallas;
+
+pub struct ZsaNullifierParams {
+    pub cond_swap_chip: CondSwapChip<pallas::Base>,
+    pub split_flag: AssignedCell<pallas::Base, pallas::Base>,
```

Comment:
To clarify their actual publicity:
```suggestion
pub(super) struct ZsaNullifierParams {
    pub(super) cond_swap_chip: CondSwapChip<pallas::Base>,
    pub(super) split_flag: AssignedCell<pallas::Base, pallas::Base>,
```